pkg/controllers/hostedcluster/controller/controller.go

```diff
@@ -0,0 +1,215 @@
+	operatorclient "open-cluster-management.io/api/client/operator/clientset/versioned"
+	operatorapiv1 "open-cluster-management.io/api/operator/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/stolostron/multicluster-controlplane/pkg/controllers/klusterlet/helpers"
+)
+
+const (
+	// hostedClusterImportAnnotation indicates whether to import hostedcluster as managed cluster
+	hostedClusterImportAnnotation = "cluster.open-cluster-management.io/import-managedcluster"
+	// hostedClusterImportFinalizer is the finalizer added to hostedcluster for importing as managed cluster
+	hostedClusterImportFinalizer = "cluster.open-cluster-management.io/import-managedcluster-finalizer"
+)
+
+// HostedClusterReconciler reconciles a HostedCluster object
+type HostedClusterReconciler struct {
+	client.Client
+	// Scheme *runtime.Scheme
+	ControlplaneKubeClient     kubernetes.Interface
+	ControlplaneOperatorClient operatorclient.Interface
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *HostedClusterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+	log.V(1).Info("Reconciling hostecluster", "namespace", req.NamespacedName.Namespace, "name", req.NamespacedName.Name)
+
+	hostedCluster := &hyperv1beta1.HostedCluster{}
+	if err := r.Get(ctx, req.NamespacedName, hostedCluster); err != nil {
+		log.Error(err, "unable to fetch hostedcluster")
+		// we'll ignore not-found errors, since they can't be fixed by an immediate
+		// requeue (we'll need to wait for a new notification), and we can get them
+		// on deleted requests.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// hostedcluster is terminating, detach the managed cluster and return
+	if !hostedCluster.GetDeletionTimestamp().IsZero() {
+		if containStr(hostedCluster.GetFinalizers(), hostedClusterImportFinalizer) {
+			if err := r.pruneManagedClusterResources(ctx, log, hostedCluster); err != nil {
+				return ctrl.Result{}, err
+			}
+			hostedCluster.SetFinalizers(removeStr(hostedCluster.GetFinalizers(), hostedClusterImportFinalizer))
+			if err := r.Client.Update(ctx, hostedCluster, &client.UpdateOptions{}); err != nil {
+				if errors.IsConflict(err) {
+					log.Info("conflict when removing finalizer from hostedcluster instance", "hostedcluster", req.NamespacedName)
+					return ctrl.Result{Requeue: true}, nil
+				} else if err != nil {
+					log.Error(err, "unable to remove finalizer to hostedcluster instance", "hostedcluster", req.NamespacedName)
+					return ctrl.Result{}, err
+				}
+			}
+		}
+
+		return ctrl.Result{}, nil
+	}
+
```

Comment:
Does the `HostedCluster ` have some condition to indicate the hosted cluster is ready to work? if it has, I think we need to wait the condition

Reply:
find the condition in the predication